### PR TITLE
Add Debounce to filter selection

### LINKED ISSUE #505: Debounce filter selection
## Problem
If you select several filters one after another it will execute one query per filter selected (even if the query takes so long that you never end up seeing it before you select another filter).

## Description

We can debounce executing the query so that it'll only execute a query after some number of milliseconds of no filters being selected. This would prevent any unnecessary queries being executed.

## Alternatives
Require users to click some sort of "apply" button to update filters rather than updating them immediately when they change.

## Implementation

- [ ] 🙋 I would be interested in implementing this feature.

### PR DESCRIPTION
**Pull request by anton202**
## Fixes

Fixes #505 by @sarayourfriend 

## Description
as described in the issue by @sarayourfriend, currently when selecting several filters one after the other
a query will be excuted for each filter selection.
so, as suggested in the issue i added a debounce of 500ms to filter selection (basically switched the watch function to watchDebounced )

<b>Staging</b>

https://user-images.githubusercontent.com/29175506/213170688-53d17a10-0bb6-4039-88e6-88bd50cfa0bf.mp4

<b>This PR</b>

https://user-images.githubusercontent.com/29175506/213171268-41363f93-a314-4d4f-8644-e3d18fb12bb7.mp4

## Testing Instructions

follow the steps in the video :)

## Checklist

- [x] My pull request has a descriptive title (not a vague title like
      `Update index.md`).
- [x] My pull request targets the _default_ branch of the repository (`main`) or
      a parent feature branch.
- [x] My commit messages follow [best practices][best_practices].
- [x] My code follows the established code style of the repository.
- [x] I added or updated tests for the changes I made (if applicable).
- [ ] I added or updated documentation (if applicable).
- [x] I tried running the project locally and verified that there are no visible
      errors.

[best_practices]:
  https://git-scm.com/book/en/v2/Distributed-Git-Contributing-to-a-Project#_commit_guidelines

## Developer Certificate of Origin

<details>
<summary>Developer Certificate of Origin</summary>

```
Developer Certificate of Origin
Version 1.1

Copyright (C) 2004, 2006 The Linux Foundation and its contributors.
1 Letterman Drive
Suite D4700
San Francisco, CA, 94129

Everyone is permitted to copy and distribute verbatim copies of this
license document, but changing it is not allowed.


Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
```

</details>

### CODE REVIEW

**Comment by anton202** [comment]
Hi, @zackkrida 
i`m having a wierd thing with e2e/mobile-menu-old.spec.ts test.
currently it fails, it fails also when i run it locally.
but, if i set the new_header:{status:{staging:disabled}} at features-flags.json
the test dose pass.
maybe its something related to ` enableOldHeader` function at the e2e/mobile-menu-old.spec.ts test ?

**Comment by sarayourfriend** [comment]
@obulat would probably know best how to debug this issue quickly :crossed_fingers:

**Comment by obulat** [comment]
@anton202, I think you opened this PR before the `new_header` flag was set to be `on` by default in the staging environment. So, in your branch, `new_header.status.staging` in `features-flags.json` is set to `switchable`, and the `defaultValue` is `off`, but in `main` it was changed to `on`. You will probably need to rebase onto (updated) main branch to fix the problems.

If there are still problems with the e2e remaining, I think we will also need to change the `enableOldHeader` function, just like you mentioned.

I tried rebasing your branch onto main, and changing the functions to what's listed below, and didn't get any errors in e2e tests locally.  

```
--- a/test/playwright/utils/navigation.ts
+++ b/test/playwright/utils/navigation.ts
@@ -464,19 +464,12 @@ export const enableNewHeader = async (page: Page) => {
 
 export const enableOldHeader = async (page: Page) => {
   // Add the new_header cookie
-  await page.context().addCookies([
-    {
-      name: "features",
-      value: "%7B%22new_header%22%3A%22off%22%7D",
-      domain: "localhost",
-      path: "/",
-    },
-  ])
+  await setCookies(page.context(), { features: { new_header: "off" } })
 }
 
 export const setCookies = async (
   context: BrowserContext,
-  cookies: Record<string, string | boolean | string[]>
+  cookies: Record<string, string | boolean | string[] | Record<string, string>>
 ) => {
   await context.addCookies(
     Object.entries(cookies).map(([name, value]) => ({
```

**Review by obulat — CHANGES_REQUESTED** [review]
Thank you for your fix, @anton202, it works perfectly! I've added notes about the tests, and I'll approve after they are fixed.

**Comment by anton202** [comment]
@obulat Thank you for the help. i`ll adress that on sunday :)

File: test/playwright/e2e/search-navigation-old.spec.ts [hunk 1]
```diff
@@ -24,6 +24,9 @@ test.describe("search history navigation", () => {
 
     // Apply a filter
     await page.click("#modification")
+    // There is a debounce of 500 ms when choosing a filter.
+    //we need to wait for the page to get reloaded before procceding the test
```
Comment (zackkrida) [inline test/playwright/e2e/search-navigation-old.spec.ts:28]:
```suggestion
    // we need to wait for the page to reload before running the test
```

here is a small grammatical suggestion

File: test/playwright/e2e/search-navigation.spec.ts [hunk 2]
```diff
@@ -28,6 +28,9 @@ test.describe("search history navigation", () => {
 
       // Apply a filter
       await page.click("#modification")
+      // There is a debounce of 500 ms when choosing a filter.
+      //we need to wait for the page to get reloaded before procceding the test
```
Comment (zackkrida) [inline test/playwright/e2e/search-navigation.spec.ts:32]:
```suggestion
      // we need to wait for the page to reload before running the test
```

File: src/components/VFilters/VSearchGridFilter.vue [hunk 3]
```diff
@@ -115,13 +116,14 @@ export default defineComponent({
...
-    watch(
+    watchDebounced(
       () => searchStore.searchQueryParams,
       (newQuery: ApiQueryParams, oldQuery: ApiQueryParams) => {
         if (!areQueriesEqual(newQuery, oldQuery)) {
           router.push(searchStore.getSearchPath())
         }
-      }
+      },
+      { debounce: 500, maxWait: 5000 }
```
Comment (zackkrida) [inline src/components/VFilters/VSearchGridFilter.vue:126]: After playing with this a little, I think something like 1000 or 800 ms for the 	`debounce` value might work better and cover more cases, including situations where a user is selecting multiple filters in a row. For example, choosing multiple sources.

**Comment by anton202** [comment]
@obulat i\`ve tried your suggestion but the test still didnt pass :(
i did noticed though that on this branch and on the main branch `new_header.status.staging` is set to `enabled`
i changed it to `switchable` and now the test passing.
btw, right now we can\`t toggle betwen the new and old header on staging. i think beacuse `new_header.status.staging` is set to  `enabled`

**Comment by anton202** [comment]
weird. all the tests pass when i run them locally
debounce-tests

**Comment by sarayourfriend** [comment]
@anton202 Sorry for the delay in responses on this PR, folks have been travelling the past week and doing work offline. Most will be back later in the week and someone will be able to help debug more closely. Appreciate the patience :pray:

**Comment by anton202** [comment]
sure, no problem :)

**Review by obulat — APPROVED** [review]

File: src/components/VFilters/VSearchGridFilter.vue [hunk 4]
```diff
@@ -115,13 +116,14 @@ export default defineComponent({
...
-    watch(
+    watchDebounced(
       () => searchStore.searchQueryParams,
       (newQuery: ApiQueryParams, oldQuery: ApiQueryParams) => {
         if (!areQueriesEqual(newQuery, oldQuery)) {
           router.push(searchStore.getSearchPath())
         }
-      }
+      },
+      { debounce: 800, maxWait: 5000 }
```
Comment (dhruvkb) [inline src/components/VFilters/VSearchGridFilter.vue:126]: What is the reasoning behind these values?

**Review by dhruvkb — APPROVED** [review]
LGTM and it works. But I have a question about the wait numbers. A code comment `//` explaining them would be good.

File: src/components/VFilters/VSearchGridFilter.vue (hunk 4 above)
Comment (dhruvkb) [reply src/components/VFilters/VSearchGridFilter.vue:126]: Ah I see @zackkrida wanted to raise it from 500. 800 seemed a very specific, very arbitrary choice to start with.

**Pull request merged**